### ISSUE TITLE
Transaction may read staled data

### ISSUE THREAD

**Issue by zyguan** [tikv/tikv#10570]
### What version of TiKV are you using?

master

### What operating system and CPU are you using?

k8s

### Steps to reproduce

Jepsen append workload still failed with incompatible-order anomaly after #10551 . It's turned out to be a new issue (see https://github.com/tikv/client-go/pull/233) and can be reproduced by the following steps.

```sql
/* init */ drop table if exists t;
/* init */ create table t (id int primary key, v int);
/* init */ insert into test (id, v) values (1, 1);
/* t1 */ set tidb_enable_async_commit = 0;
/* t1 */ set tidb_enable_1pc = 0;
/* t2 */ set tidb_enable_async_commit = 0;
/* t2 */ set tidb_enable_1pc = 0;
/* t1:wait */ begin optimistic;
/* t1 */ insert into tk values (1, 2) on duplicate key update val = val + 10;
/* t2:wait */ begin pessimistic;
-- failpoint.Enable("tikvclient/beforeCommit", "1*pause")
/* t1 */ commit;
-- wait a while until t1 prewrites
/* t2 */ insert into tk values (1, 3) on duplicate key update val = val + 100;
-- blocked, wait a while until t2 starts adding pessimistic locks
-- failpoint.Disable("tikvclient/beforeCommit")
/* t2:wait */ commit;
/* t2 */ select * from test where id = 1; -- (should be 111, but get 101)
```

### What did you expect?

As described in *Steps to reproduce*.

### What did happened?

As described in *Steps to reproduce*.
